### ISSUE TITLE
milestone: reproduce ResNet-18/CIFAR-10 (~94–95%) fully in-GUI with fixed seed — the teaching-tool → research-tool watershed

### ISSUE THREAD

**Issue by treeleaves30760** [treeleaves30760/CodefyUI#138]
> 中文摘要：分水嶺驗收：純 GUI 配置、seed 固定，重現 ResNet-18/CIFAR-10 標準 baseline（~94–95%），全程 Run Service + 曲線落地。

## Context
This is a validation milestone, not a feature: it proves gaps #1–#7 are actually closed. Standard recipe: ResNet-18 (CIFAR variant: 3×3 stem, no maxpool), SGD momentum 0.9, lr 0.1, cosine or step schedule, weight decay 5e-4, batch 128, ~100–200 epochs, RandomCrop(32,4)+Flip+CIFAR normalize → 94–95% test accuracy.

## Spec
1. Build the graph in the GUI using: existing ResNet preset (verify the CIFAR stem variant exists — if the preset is ImageNet-style, add a CIFAR variant preset), P4-21 augmentation chain, P4-19 full params (momentum/weight_decay/label_smoothing optional), P4-20 (optionally bf16), LRScheduler cosine, seed fixed.
2. Run via Run Service on the RTX 5090 dogfood machine: queue it, close the browser, come back — the whole MVP story exercised end-to-end.
3. Deliverables:
   - `examples/Usage_Example/ResNet18-CIFAR10-Baseline/` (graph.json + README documenting expected accuracy, wall-clock, seed, and the exact recipe provenance).
   - Docs page `docs/docs/usage/reproducing-baselines.md` — the "research-grade walkthrough".
   - Metrics CSV + loss/accuracy curves attached to the issue as the acceptance evidence.
   - CI smoke variant: 2-epoch run asserting the pipeline executes and accuracy > random (fast, CPU-or-GPU).
4. File a bug for every friction point encountered (this milestone is also a structured dogfood audit — expect it to spawn issues).

## Acceptance Criteria
- [ ] Test accuracy ≥ 94.0% with the documented seed, achieved without writing any Python outside the GUI (PythonScript node allowed only for metrics glue if truly needed — document if used).
- [ ] Re-running with the same seed reproduces final accuracy within ±0.2pp.
- [ ] The run survived at least one deliberate browser disconnect + one Stop→resume cycle during development (exercise P1-06/07/P0-03 in anger).
- [ ] Example ships and is executed by the examples CI (`test_builtin_examples.py` pattern, short-epoch mode).

## Tests
- The CI smoke example; otherwise this issue's "test" is the documented evidence bundle.

## Dependencies
Depends on P4-19, P4-20, P4-21 (and Wave 1 throughout). Closes the "research-grade" claim for training.

**Comment by treeleaves30760** [comment]
## Acceptance evidence

PR #206. Measured on this machine — note it is an **RTX 4080**, not the RTX 5090 the issue text assumes, so the wall clock below is the real number for the real hardware.

### Headline

**95.48% top-1 test accuracy on CIFAR-10**, 200 epochs, 22 min 29 s, configured entirely from GUI nodes.

| | |
|---|---|
| Hardware | NVIDIA GeForce RTX 4080 (16 GB, compute capability 8.9) |
| Software | PyTorch 2.11.0+cu128, torchvision 0.26.0, Python 3.11, Windows 11 |
| Run options | `device: cuda`, `seed: 1337`, `deterministic: true` |
| Accuracy | **95.48%** (9,548 / 10,000) |
| Wall clock | 1,349.3 s, 6.62 s/epoch mean (min 6.20, max 8.14) |
| Final train loss | 0.001669 |
| Final test loss | 0.172160 |

Comfortably inside the 94-95% the issue predicted.

### Criterion 1 — at or above 94.0%, no Python outside the GUI

Met. The example ships `graph.json`, `README.md` and the evidence files. There is no `model.py` and no `PythonScript` node — verified by inspecting the node-type census of the shipped graph:

```
CheckpointSaver 1, DataLoader 2, Dataset 2, EvaluateModel 1, LRScheduler 1,
Loss 1, NormalizeTransform 2, Optimizer 1, RandomCrop 1, RandomHorizontalFlip 1,
SequentialModel 1, Start 1, ToTensorTransform 2, TrainingLoop 1, Visualize 1
```

The architecture is 70 layers inside the `SequentialModel` layer editor. It is the **CIFAR variant** — 3x3 stride-1 stem, no maxpool — and its parameter count, **11,173,962**, matches `torchvision.models.resnet18(num_classes=10)` with the stem swapped and the maxpool removed, exactly.

On the preset question in spec item 1: there is no ResNet node and no ResNet preset in the repo at all. The two existing "ResNet" artifacts both use the **ImageNet** stem (7x7 stride-2 plus maxpool) — `examples/Usage_Example/ResNet-CIFAR10/TrainResNet-CIFAR10` and `examples/Model_Architecture/ResNet-SkipConnection-CNN`. The only correct CIFAR stem in the tree was reference Python in a `model.py`, not anything a node could use. The new example is the first CIFAR-correct ResNet expressed as nodes.

### Criterion 2 — same seed within 0.2pp

Met, with room to spare. The second run was submitted **after a server restart**, so it shared no process state with the first.

| | |
|---|---|
| Run A | 95.48% |
| Run B | 95.48% |
| Difference | **0.00 pp** |
| Metric points compared | 601 (200 train loss, 200 test loss, 200 lr, 1 accuracy) |
| Points differing | **0** — bitwise identical throughout |

Bitwise equality on CUDA follows from `deterministic: true` disabling cuDNN's autotuner, which costs about 4% of throughput here. This is #134 cashing in.

### Criterion 3 — browser disconnect and stop-then-resume

Both met, separately.

**Disconnect.** The browser tab was closed at epoch 4 of the 200-epoch run. The run continued server-side to epoch 118 with nothing connected. A brand-new tab then re-attached and showed the live run at 13m 56s elapsed with a current loss, alongside the second run still sitting at queue position 1.

**Stop and resume**, on the same recipe at 20 epochs (seed 2024):

1. Stopped at epoch 8. The cooperative stop wrote an interrupt checkpoint automatically: `interrupted/…-train-e8b272.pt`, meta `{epoch: 8, batch: 272}`, registered as a run artifact (alongside a TensorBoard artifact).
2. A resumed graph with `CheckpointLoader` wired to `TrainingLoop.start_epoch` logged **exactly epochs 9 through 20** and finished at 91.83%.
3. The learning rate at every resumed epoch matched the theoretical cosine from `base_lrs` 0.1 with `T_max` 20 to within 1e-12 — the schedule was continuous across the stop, not restarted.

Point 3 is the interesting one, because it is where **#149** lives. It did not bite, because the graph wires `Optimizer` into `LRScheduler` into `CheckpointLoader`. Wired the other way round — loader first — the scheduler is constructed from an optimizer whose learning rate has already been restored to a decayed value, and captures that as `base_lrs`. Both docs pages now carry a warning admonition about the ordering.

### Criterion 4 — executed by the examples CI

Met. `test_resnet18_cifar10_baseline_short_epoch` in `backend/tests/test_builtin_examples.py` executes the shipped graph for two optimizer steps in about 3 seconds, with no download.

The generic parametrised test only *validates* this graph, because it contains `Dataset`/`TrainingLoop` and is on the slow list. That is not enough here: the 70-node layer spec is an opaque JSON string to `validate_graph`, so the thing most likely to rot is precisely the thing structural validation cannot see. The smoke swaps the two `Dataset` nodes for `ImageFolderDataset` against a generated image folder — a drop-in, same transform ports, same `dataset` output — so every edge, the augmentation chain and the whole layer spec execute as shipped. It then asserts the 11,173,962 parameter count, a finite loss after exactly two steps, evaluation over the held-out split, and the checkpoint file appearing where `CheckpointSaver` said.

### Evidence bundle

Committed next to the example rather than attached here, so it stays version-controlled:

- `examples/Usage_Example/ResNet18-CIFAR10-Baseline/evidence/metrics-seed1337.csv` — the full metric export, exactly as `GET /api/runs/<id>/metrics?format=csv` produced it
- `examples/Usage_Example/ResNet18-CIFAR10-Baseline/evidence/curves-seed1337.png` — loss curves and the cosine schedule, plotted from that CSV

The bundle has **no accuracy-against-epoch curve**, which the issue asked for, because no node can emit one — see #202 below.

### Issues this milestone spawned

As expected of a structured dogfood audit, five:

| | |
|---|---|
| #201 | `validate_graph` accepts a graph `execute_graph` refuses, when a root node has no trigger edge. Minimal 3-node repro in the issue. This is the one that cost real time — the graph validated clean and the run then failed naming a node far downstream. |
| #202 | No per-epoch accuracy metric anywhere. `TrainingLoop` records loss and lr; `EvaluateModel` records a single point at step 1. Directly why the evidence above is missing its accuracy curve. |
| #203 | No periodic checkpointing. Hit for real — a server process death took out 101 completed epochs with zero recoverable artifacts. Survivable at 22 minutes, not at 12 hours. |
| #204 | `EvaluateModel.device` has no `auto` and defaults to `cpu`, silently ignoring the run device. |
| #205 | `LRScheduler.T_max` must equal `TrainingLoop.epochs`, nothing says so, and getting it wrong degrades accuracy silently. |

None are fixed in #206 — a validation milestone should not also be changing engine semantics — and each is either worked around in the example or documented on the docs page.

### Docs

`docs/docs/usage/reproducing-baselines.md` plus the zh-TW twin: the recipe, why the CIFAR stem differs from torchvision's and what it costs to get wrong, how `seed` and `deterministic` interact, the measured numbers, the disconnect story, the stop/resume wiring with the #149 ordering warning, and the gotchas above.
